File: tests/r_parity/20_bacon.py
"""StatsPAI Goodman-Bacon decomposition parity (Python side) -- Module 20.

Runs sp.bacon_decomposition on the mpdta replica. The companion
20_bacon.R uses bacondecomp::bacon. Tolerance: rel < 1e-3 on the
overall TWFE coefficient.

The decomposition returns per-cohort-pair comparisons; we compare
the overall TWFE coefficient (a scalar that both sides should agree
on) plus the per-cohort-pair point estimates indexed by
(treated_cohort, comparison_cohort).
"""
from __future__ import annotations

import statspai as sp

from _common import ParityRecord, dump_csv, write_results


MODULE = "20_bacon"


def main() -> None:
    df = sp.datasets.mpdta()
    dump_csv(df, MODULE)

    res = sp.bacon_decomposition(
        df, y="lemp", treat="treat", time="year", id="countyreal"
    )

    rows: list[ParityRecord] = [
        ParityRecord(
            module=MODULE, side="py", statistic="beta_twfe",
            estimate=float(res["beta_twfe"]),
            n=int(len(df)),
        ),
        ParityRecord(
            module=MODULE, side="py", statistic="weighted_sum",
            estimate=float(res["weighted_sum"]),
            n=int(len(df)),
        ),
        ParityRecord(
            module=MODULE, side="py", statistic="negative_weight_share",
            estimate=float(res["negative_weight_share"]),
            n=int(len(df)),
        ),
    ]

    # Per-cohort-pair point estimates.
    decomp = res["decomposition"]
    for _, row in decomp.iterrows():
        treated = int(row["treated"])
        control = row["control"]
        if isinstance(control, str):
            control_str = "never"
        else:
            control_str = str(int(control))
        rows.append(
            ParityRecord(
                module=MODULE, side="py",
                statistic=f"pair_{treated}_vs_{control_str}_est",
                estimate=float(row["estimate"]),
                n=int(len(df)),
            )
        )

    write_results(
        MODULE, "py", rows,
        extra={
            "n_comparisons": int(res["n_comparisons"]),
            "already_treated_control_weight_share": (
                float(res["already_treated_control_weight_share"])
            ),
            "decomposition_note": (
                "StatsPAI follows R/Stata bacondecomp dyad windows "
                "and weighting for the uncontrolled Goodman-Bacon "
                "decomposition. The overall TWFE coefficient, "
                "weighted sum, negative-weight mass, and pairwise "
                "2x2 estimates match the R reference to numerical "
                "precision; Stata's ado exports only the aggregate "
                "rows in the committed golden artifact."
            ),
        },
    )


if __name__ == "__main__":
    main()

File: tests/r_parity/20_bacon.R
# StatsPAI Goodman-Bacon decomposition parity (R side) -- Module 20.
#
# Reads data/20_bacon.csv (the StatsPAI mpdta replica) and runs
# bacondecomp::bacon. Tolerance: rel < 1e-3 on the overall TWFE
# coefficient.

.args <- commandArgs(trailingOnly = FALSE)
.file_arg <- grep("^--file=", .args, value = TRUE)
.script_dir <- if (length(.file_arg) > 0) {
  dirname(normalizePath(sub("^--file=", "", .file_arg[1])))
} else {
  getwd()
}
source(file.path(.script_dir, "_common.R"))

suppressPackageStartupMessages({
  library(bacondecomp)
})

MODULE <- "20_bacon"

df <- read_csv_strict(MODULE)

# bacondecomp::bacon returns either a tibble of (type, weight,
# avg_est) -- when summary_only is TRUE -- or a list of two tibbles
# when summary_only is FALSE. Default behaviour returns a list and
# prints the summary.
res <- bacondecomp::bacon(lemp ~ treat, data = df,
                          id_var = "countyreal", time_var = "year",
                          quietly = TRUE)

# `res` is the per-pair data.frame.
beta_twfe <- sum(res$weight * res$estimate)
weighted_sum <- beta_twfe   # same identity
neg_weight_share <- sum(res$weight[res$weight < 0]) / sum(abs(res$weight))
n_comparisons <- nrow(res)

rows <- list(
  parity_row(MODULE, "beta_twfe", estimate = beta_twfe, n = nrow(df)),
  parity_row(MODULE, "weighted_sum", estimate = weighted_sum, n = nrow(df)),
  parity_row(MODULE, "negative_weight_share",
             estimate = neg_weight_share, n = nrow(df))
)

# Per-pair estimates. R uses 99999 to denote untreated.
for (i in seq_len(nrow(res))) {
  treated_cohort <- as.integer(res$treated[i])
  ctrl_raw <- as.integer(res$untreated[i])
  ctrl_str <- if (ctrl_raw == 99999) "never" else as.character(ctrl_raw)
  rows[[length(rows) + 1L]] <- parity_row(
    module    = MODULE,
    statistic = paste0("pair_", treated_cohort, "_vs_", ctrl_str, "_est"),
    estimate  = res$estimate[i],
    n         = nrow(df)
  )
}

write_results(MODULE, rows, extra = list(n_comparisons = n_comparisons))
